pkg/controller/policies_synthesizer.go

```diff
@@ -0,0 +1,71 @@
+package controller
+
+import (
+	networking "k8s.io/api/networking/v1"
+
+	"github.com/np-guard/cluster-topology-analyzer/pkg/common"
+)
+
+type PoliciesSynthesizer struct {
+	logger      Logger
+	stopOnError bool
+
+	errors []FileProcessingError
+}
+
+type PoliciesSynthesizerOption func(*PoliciesSynthesizer)
+
+func WithLogger(logger Logger) PoliciesSynthesizerOption {
+	return func(p *PoliciesSynthesizer) {
+		p.logger = logger
+	}
+}
+
+func WithStopOnError() PoliciesSynthesizerOption {
+	return func(p *PoliciesSynthesizer) {
+		p.stopOnError = true
+	}
+}
+
+func NewPoliciesSynthesizer(options ...PoliciesSynthesizerOption) *PoliciesSynthesizer {
+	// object with default behavior options
+	ps := &PoliciesSynthesizer{
+		logger:      NewDefaultLogger(),
+		stopOnError: false,
+		errors:      []FileProcessingError{},
+	}
+	for _, o := range options {
+		o(ps)
+	}
+	return ps
+}
+
+func (ps *PoliciesSynthesizer) Errors() []FileProcessingError {
+	return ps.errors
+}
+
+func (ps *PoliciesSynthesizer) PoliciesFromFolderPath(dirPath string) ([]*networking.NetworkPolicy, error) {
+	activeLogger = ps.logger
+
+	emptyStr := ""
+	args := common.InArgs{}
+	args.DirPath = &dirPath
+	args.CommitID = &emptyStr
+	args.GitBranch = &emptyStr
+	args.GitURL = &emptyStr
+
+	connections, errs := extractConnections(args, ps.stopOnError)
+	policies := []*networking.NetworkPolicy{}
+	if !returnOn1StError(ps.stopOnError, errs) {
+		policies = synthNetpols(connections)
+	}
+
+	ps.errors = errs
+	for _, err := range errs {
+		if err.IsFatal() {
+			return nil, err.Error()
+		}
+	}
```

Comment:
This is fine and it implements the agreed interface. But it sparkles some thoughts about potential confusion source.

I though that we can guarantee by definition that there is always 0 or 1 fatal errors (max 1, as seeing one fatal error forces us to stop processing). What we see here is our classification of the severity `fatal` which means that the lib could theoretically continue work till the end if the user has chosen so.

I think that there are the following type of errors here:
1. `fatal` where the processing must stop now - e.g., file not found, not enough memory
2. The severity that we called fatal = the result is garbage
3. The severity that we called critical = the result may be incomplete

Now, using word fatal for 1 and 2 makes it confusing. Maybe we could reiterate the naming?

In line 66, we return the second fatal, whereas when I proposed it I was thinking about the first fatal.
Let me know what you think.

Reply:
We don't really have errors that are of type 1 but not type 2. i.e., whenever we have to stop immediately, the results cannot be used. I cannot really imagine a case where the results (NetworkPolicies) will be meaningful if we ran out of memory. I prefer to be on the safe side here.